CommonCompilerConfig.cmake

```diff
@@ -0,0 +1,239 @@
+  set(COMMON_COMPILER_DEFINITIONS
+      ${COMMON_COMPILER_DEFINITIONS}
+      -DHAVE_UNISTD_H)
+elseif(MSVC OR MSYS OR MINGW)
+  set(COMMON_COMPILER_DEFINITIONS
+      ${COMMON_COMPILER_DEFINITIONS}
+      -DPSAPI_VERSION=1
+      -DWIN32
+      -D_WINDOWS
+      -D_MBCS
+      -D_CRT_SECURE_NO_WARNINGS
+      -DNDEBUG
+      -D_VARIADIC_MAX=10
+      -DNOMINMAX)
+  if(MSYS OR MINGW)
+    set(COMMON_COMPILER_DEFINITIONS
+        ${COMMON_COMPILER_DEFINITIONS}
+        -DHAVE_UNISTD_H)
+  endif()
+endif()
+
+
+#
+# Set linker (ld)
+# use ld.gold if available
+#
+execute_process(COMMAND ld.gold --version RESULT_VARIABLE EXIT_CODE)
+if(EXIT_CODE EQUAL 0)
+  message("Using ld.gold as LINKER.")
+  set(CMAKE_LINKER "ld.gold")
+endif()
+
+
+#
+# Determine stdlib settings
+#
+set(stdlib_cxx "")
+set(stdlib_common "")
+
+if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+  set(stdlib_cxx "${stdlib_cxx} -stdlib=libc++")
+endif()
+
+if (${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU")
+   set(stdlib_common "${stdlib_common} -static-libgcc")
+   set(stdlib_cxx "${stdlib_cxx} -static-libstdc++")
+endif()
+
+
+#
+# Determine Optimization flags here
+# These are quite aggresive flags, if your code misbehaves for strange reasons,
+# try compiling without them.
+#
+if(NOT ${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
+  set(optimization_flags_cc "${optimization_flags_cc} -mtune=generic -O2")
+  set(optimization_flags_cc "-pipe ${optimization_flags_cc}") #TODO use -Ofast instead of -O3
+  set(optimization_flags_lt "-O2 ${optimization_flags_lt}")
+
+  if(${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU" AND NOT MINGW)
+    set(optimization_flags_cc "${optimization_flags_cc} -fuse-ld=gold")
+    # NOTE -flto must go together in both cc and ld flags; also, it's presently incompatible
+    # with the -g option in at least some GNU compilers (saw in `man gcc` on Ubuntu)
+    set(optimization_flags_cc "${optimization_flags_cc} -fuse-linker-plugin -flto-report -flto") #TODO fix LTO for clang
+    set(optimization_flags_lt "${optimization_flags_lt} -flto") #TODO LTO for clang too
+  endif()
+endif()
+
+
+#
+# compiler specific settings here
+#
+
+if(${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
+  # MS Visual C
+  set(shared_compile_flags "/TP /Zc:wchar_t /Gm- /fp:precise /errorReport:prompt /W1 /WX- /GR /Gd /GS /Oy- /EHs /analyze- /nologo")
+  set(COMMON_LINKER_FLAGS_UNOPTIMIZED "/NOLOGO /SAFESEH:NO /NODEFAULTLIB:LIBCMT")
+  if("${BITNESS}" STREQUAL "32")
+    set(COMMON_LINKER_FLAGS_UNOPTIMIZED "${COMMON_LINKER_FLAGS_UNOPTIMIZED} /MACHINE:X86")
+  else()
+    set(COMMON_LINKER_FLAGS_UNOPTIMIZED "${COMMON_LINKER_FLAGS_UNOPTIMIZED} /MACHINE:X${BITNESS}")
+  endif()
+
+else()
+  # LLVM Clang / Gnu GCC
+  set(COMMON_CXX_FLAGS_UNOPTIMIZED "${COMMON_CXX_FLAGS_UNOPTIMIZED} ${stdlib_cxx} -std=c++11")
+
+  set(shared_compile_flags "-m${BITNESS} ${stdlib_common} -Wextra -Wreturn-type -Wunused -Wno-unused-variable -Wno-unused-parameter -Wno-missing-field-initializers")
```

Comment:
if `shared_compile_flags` are the flags for building libraries, we probably could drop all the `W*` flags.